service/worker/failoverManager/workflow.go

```diff
@@ -0,0 +1,376 @@
+		Domains       []string
+		TargetCluster string
+	}
+
+	// FailoverActivityResult result for failover activity
+	FailoverActivityResult struct {
+		SuccessDomains []string
+		FailedDomains  []string
+	}
+
+	// QueryResult ..
+	QueryResult struct {
+		TotalDomains   int
+		Success        int
+		Failed         int
+		State          string
+		TargetCluster  string
+		SourceCluster  string
+		SuccessDomains []string
+		FailedDomains  []string
+	}
+)
+
+func init() {
+	workflow.RegisterWithOptions(FailoverWorkflow, workflow.RegisterOptions{Name: WorkflowTypeName})
+	activity.RegisterWithOptions(FailoverActivity, activity.RegisterOptions{Name: failoverActivityName})
+	activity.RegisterWithOptions(GetDomainsActivity, activity.RegisterOptions{Name: getDomainsActivityName})
+}
+
+// FailoverWorkflow is the workflow that managed failover all domains with IsManagedByCadence=true
+func FailoverWorkflow(ctx workflow.Context, params *FailoverParams) (*FailoverResult, error) {
+	err := validateParams(params)
+	if err != nil {
+		return nil, err
+	}
+
+	// define query properties
+	var failedDomains []string
+	var successDomains []string
+	var totalNumOfDomains int
+	wfState := wfRunning
+	err = workflow.SetQueryHandler(ctx, QueryType, func(input []byte) (*QueryResult, error) {
+		return &QueryResult{
+			TotalDomains:   totalNumOfDomains,
+			Success:        len(successDomains),
+			Failed:         len(failedDomains),
+			State:          wfState,
+			TargetCluster:  params.TargetCluster,
+			SourceCluster:  params.SourceCluster,
+			SuccessDomains: successDomains,
+			FailedDomains:  failedDomains,
+		}, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// get target domains
+	ao := workflow.WithActivityOptions(ctx, getGetDomainsActivityOptions())
+	getDomainsParams := &GetDomainsActivityParams{
+		TargetCluster: params.TargetCluster,
+		SourceCluster: params.SourceCluster,
+		Domains:       params.Domains,
+	}
+	var domains []string
+	err = workflow.ExecuteActivity(ao, GetDomainsActivity, getDomainsParams).Get(ctx, &domains)
+	if err != nil {
+		return nil, err
+	}
+	totalNumOfDomains = len(domains)
+
+	// failover in batch
+	ao = workflow.WithActivityOptions(ctx, getFailoverActivityOptions())
+	batchSize := params.BatchFailoverSize
+	times := len(domains) / batchSize
+
+	pauseCh := workflow.GetSignalChannel(ctx, PauseSignal)
+	resumeCh := workflow.GetSignalChannel(ctx, ResumeSignal)
+	var shouldPause bool
+
+	for i := 0; i < times; i++ {
+		// check if need to pause
+		shouldPause = pauseCh.ReceiveAsync(nil)
+		if shouldPause {
+			wfState = wfPaused
+			resumeCh.Receive(ctx, nil)
+		}
+		wfState = wfRunning
+
+		failoverActivityParams := &FailoverActivityParams{
+			Domains:       domains[i*batchSize : min((i+1)*batchSize, totalNumOfDomains)],
+			TargetCluster: params.TargetCluster,
+		}
+		var actResult FailoverActivityResult
+		workflow.ExecuteActivity(ao, FailoverActivity, failoverActivityParams).Get(ctx, &actResult)
```

Comment:
I think we need to properly handle the error here. If worker dies while executing the failover activity, we don't even know which domains have succeeded and which have failed. This will also be a problem for rollback as we don't have the accurate information for successDomains.

Reply:
good catch. updated.